DOCSP-32166 - Write Basics (#16)
mongodb/docs-entity-framework: source/fundamentals/write-data.txt

.. _entity-framework-write-data:

=====================
Write Data to MongoDB
=====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: EF, EF Core, code example, write

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

{+framework-core+} allows you to work with data in your application without

> explicitly running database commands. You can perform write operations such as
> inserting, updating, or deleting data in your application, and persist those

Reviewer: ```suggestion
explicitly running database commands. You can perform write operations to
insert, update, or delete data within your application, and then persist those
```

> changes to MongoDB using the ``SaveChanges()`` method.

Reviewer: https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#u:~:text=word%2C%20username.-,using,-by%20using%20with
```suggestion
changes to MongoDB by using the ``SaveChanges()`` method.
```

> When ``SaveChanges()`` is called, the {+provider-short+} automatically detects
> any changes made to your data and runs the necessary commands to update the
> database using the MongoDB Query API.

Reviewer: ```suggestion
When you call the ``SaveChanges()`` method, the {+provider-short+} detects
any changes made to your data and sends commands to update the
database by using the MongoDB Query API.
```

> In this guide you can see examples of common write operations on
> an application configured to use the {+provider-long+}.

Reviewer: ```suggestion
In this guide, you can see examples of how to perform common write operations on
an application configured to use the {+provider-long+}.
```

.. tip::

   To learn how to configure an application to use the {+provider-short+}, see
   :ref:`entity-framework-configure`.

Insert
------

> You can insert a single entity into your collection by using the ``Add()``
> method, or insert multiple entities at once by using the ``AddRange()`` method. 

Reviewer: S: reverse structure:
```suggestion
You can use the ``Add()`` method to insert a single entity into your collection, 
or you can use the ``AddRange()`` method to insert multiple entities at once.
```

Insert One Entity
~~~~~~~~~~~~~~~~~

> The ``Add()`` method accepts a single entity of the same type as that of the ``DbSet``
> on which it's being called.

Reviewer: ```suggestion
The ``Add()`` method accepts a single entity of the same type that you previously
specified on the ``DbSet`` instance that you are modifying.
```

The following code uses the ``Add()`` method to add a new ``Planet`` object to
the ``DbSet`` called ``Planets``. It then calls the ``SaveChanges()`` method to
insert that entity into the MongoDB collection.

.. literalinclude:: /includes/code-examples/quick-reference.cs
    :start-after: // start-insert-one
    :end-before: // end-insert-one
    :language: csharp

>     :dedent:

Reviewer: S: Applies to all code examples - reduce horizontal scroll by newlining fields

Insert Multiple Entities
~~~~~~~~~~~~~~~~~~~~~~~~

> The ``AddRange()`` method accepts an array of entities to add to the ``DbSet``. 

Reviewer: ```suggestion
The ``AddRange()`` method accepts an array of entities that you want to add to the ``DbSet``. 
```

The following code uses the ``AddRange()`` method to add an array of ``Planet``
objects to the ``DbSet`` called ``Planets``. It then calls the ``SaveChanges()``
method to insert those entities into the MongoDB collection.

.. literalinclude:: /includes/code-examples/quick-reference.cs
    :start-after: // start-insert-many
    :end-before: // end-insert-many
    :language: csharp
    :dedent:

Update
------

To update an entity, first retrieve the entity that you want to update. Then
make the changes to that entity in your application by using the conventions of

> your programming language. To complete the update, call the
> ``SaveChanges()`` method. The {+provider-short+} compares the updated entity with a

Reviewer: ```suggestion
your programming language. To save the update to MongoDB, call the
``SaveChanges()`` method. The {+provider-short+} compares the updated entity with a
```

snapshot of the entity before the change and automatically updates the collection
using the MongoDB Query API.

> The following code updates the ``name`` field of a ``Planet``, then calls the
> ``SaveChanges()`` method to persist that change to the collection.

Reviewer: ```suggestion
The following code retrieves an entity in which the ``name`` value is ``"Mercury"``, then
updates the ``name`` field and calls the ``SaveChanges()`` method to persist that change to the collection.
```

.. literalinclude:: /includes/code-examples/quick-reference.cs
    :start-after: // start-update-one
    :end-before: // end-update-one
    :language: csharp
    :dedent:

Delete
------

You can delete a single entity from your collection by using the ``Remove()``

> method, or delete multiple entities at once by using the ``RemoveRange()`` method.

Reviewer: S: same comment as earlier from the insert sectioion

Delete One Entity
~~~~~~~~~~~~~~~~~

The ``Remove()`` method accepts a single entity of the same type as that of the ``DbSet``

> on which it's being called.

Reviewer: S: same comment as earlier

> The following code removes a ``Planet`` entity with the ``name`` field of
> "Pluto". It then calls the ``SaveChanges()`` method to delete that entity from
> the MongoDB collection.

Reviewer: ```suggestion
The following code removes a ``Planet`` entity in which the ``name`` value is
``"Pluto"``. It then calls the ``SaveChanges()`` method to delete that entity from
the MongoDB collection.
```

.. literalinclude:: /includes/code-examples/quick-reference.cs
    :start-after: // start-delete-one
    :end-before: // end-delete-one
    :language: csharp
    :dedent:

Delete Multiple Entities
~~~~~~~~~~~~~~~~~~~~~~~~

The ``RemoveRange()`` method accepts an array of entities to remove from the
``DbSet``. 

> The following code finds two ``Planet`` entities and adds them to an array. It
> then calls the ``RemoveRange()`` method to remove both entities from the
> ``DbSet``. Finally, it calls the ``SaveChanges()`` method to remove those
> entities from the MongoDB collection.

Reviewer: ```suggestion
The following code finds two ``Planet`` entities and adds them to an array. It
then uses the ``RemoveRange()`` method to remove both entities from the
``DbSet``. Finally, it uses the ``SaveChanges()`` method to remove those
entities from the MongoDB collection.
```

.. _entity-framework-write-data:

=====================
Write Data to MongoDB
=====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: EF, EF Core, code example, write

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

{+framework-core+} allows you to work with data in your application without

> explicitly running database commands. You can perform write operations to

Reviewer 2: I think the "Perform write operations" sentence makes things less clear.

Author: Good call, I think this section got a little jumbled through the edits. I simplified that sentence

Insert One Entity
~~~~~~~~~~~~~~~~~

The ``Add()`` method accepts a single entity of the same type that you
specified on the ``DbSet`` instance that you are modifying.

The following code uses the ``Add()`` method to add a new ``Planet`` object to
the ``DbSet`` called ``Planets``. It then calls the ``SaveChanges()`` method to
insert that entity into the MongoDB collection.

.. literalinclude:: /includes/code-examples/quick-reference.cs
   :start-after: // start-insert-one
   :end-before: // end-insert-one
   :language: csharp
   :dedent:

Insert Multiple Entities
~~~~~~~~~~~~~~~~~~~~~~~~

The ``AddRange()`` method accepts an array of entities that you want to add to the ``DbSet``. 

The following code uses the ``AddRange()`` method to add an array of ``Planet``
objects to the ``DbSet`` called ``Planets``. It then calls the ``SaveChanges()``
method to insert those entities into the MongoDB collection.

.. literalinclude:: /includes/code-examples/quick-reference.cs
   :start-after: // start-insert-many
   :end-before: // end-insert-many
   :language: csharp
   :dedent:

Update
------

To update an entity, first retrieve the entity that you want to update. Then

> make the changes to that entity. To save the update to MongoDB, call the

Reviewer 2: I'd add "using the properties to new values or by adding and removing items from the lists or arrays the entity has."

Author: I tried to keep the "make the changes" part pretty vague so it didn't seem like we are saying that there are only certain specific types of updates supported with the provider. I think the main focus of this section should be that they have to call `SaveChanges()` after any updates they do, since that's what's specific to EF. 
Happy to workshop a little more here or through slack if you still think we should get more specific on that part!

Reviewer 2: I think we could clarify that anything changing the entity - setting properties or adding/removing items from any collections on it - will be reflected via SaveChanges thanks to the object tracking.

Author: I added a clarifying sentence, let me know if that covers what you have in mind.

.. _entity-framework-write-data:

=====================
Write Data to MongoDB
=====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: EF, EF Core, code example, write

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> {+framework-core+} allows you to work with data in your application without

Author: This constant is for "Entity Framework Core" (the framework itself) not for our provider. Since it's the product name I don't think having add "the" before would work here.

specified on the ``DbSet`` instance that you are modifying.

The following code uses the ``Add()`` method to add a new ``Planet`` object to
the ``DbSet`` called ``Planets``. It then calls the ``SaveChanges()`` method to
insert that entity into the MongoDB collection.

.. literalinclude:: /includes/code-examples/quick-reference.cs
   :start-after: // start-insert-one
   :end-before: // end-insert-one
   :language: csharp
   :dedent:

Insert Multiple Entities
~~~~~~~~~~~~~~~~~~~~~~~~

The ``AddRange()`` method accepts an array of entities that you want to add to the ``DbSet``. 

The following code uses the ``AddRange()`` method to add an array of ``Planet``
objects to the ``DbSet`` called ``Planets``. It then calls the ``SaveChanges()``
method to insert those entities into the MongoDB collection.

.. literalinclude:: /includes/code-examples/quick-reference.cs
   :start-after: // start-insert-many
   :end-before: // end-insert-many
   :language: csharp
   :dedent:

Update
------

To update an entity, first retrieve the entity that you want to update. Then
make the changes to that entity in your application by using the conventions of

> your programming language. To save the update to MongoDB, call the

Author: I think I was trying to make it clear that you update it as you normally would in any .NET/C# application, but looking back on it now I think that's an unnecessary clarification so I removed it.
